### PR TITLE
Heterogeneous memory integration

### PR DESCRIPTION
**Pull request by gcongiu**
This PR introduces NUMA-awareness and heterogeneous memory support by adding the following features to MPICH:

- [ ] `Memory locality detection`: detect NUMA nodes close to the hardware thread the current MPI process is running onto;
- [ ] `Shared memory partitioning`: allocate a separate shared memory segment (`VMA`) for every NUMA node;
- [ ] `Memory binding setting`: set memory binding for a shared memory segment to the requested memory node.

These three features combined allow MPICH to partition a shared memory object (e.g., fastboxes in point-to-point) by the number of NUMA nodes in the system. Processes will then have their memory objects bound to the closest NUMA.

REF: https://github.com/pmodels/mpich/issues/3511

### CODE REVIEW

**Comment by gcongiu** [comment]
test:jenkins/ch3/most

**Comment by gcongiu** [comment]
test:jenksins/ch3/tcp

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp

File: src/mpid/ch3/src/ch3u_win_fns.c [hunk 1]
```diff
@@ -544,3 +598,96 @@ int MPID_Win_get_info(MPIR_Win * win, MPIR_Info ** info_used)
...
+            flags |= HWLOC_MEMBIND_NOCPUBIND;
+    } else {
+        policy = HWLOC_MEMBIND_DEFAULT;
+        flags |= HWLOC_MEMBIND_NOCPUBIND;
+    }
+
+    if (MPIR_Process.mcdramset_is_valid)
+        hwloc_set_area_membind(MPIR_Process.topology, *(void **) baseptr, len,
+                               MPIR_Process.bindset_is_valid ?
+                               MPIR_Process.mcdram_sibling : MPIR_Process.mcdram_nodes,
```
Comment (gcongiu) [inline src/mpid/ch3/src/ch3u_win_fns.c]: Fixme: does not consider HWLOC_MEMBIND_INTERLEAVE

**Comment by gcongiu** [comment]
test:jenkins/ch3/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch3/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch3/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch3/most

**Comment by gcongiu** [comment]
test:jenkins/ch3/most

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp

**Comment by gcongiu** [comment]
@pavanbalaji can you start having a look at this?

**Comment by pavanbalaji** [comment]
@gcongiu This seems somewhat specific to MCDRAM and is hardcoding the buffers (fbox, cells, ..).  Can you generalize it to support placing a collection of buffers into a collection of memories?  If these can be extendable lists, that'd be great, even if they contain a fixed number of items right now.

**Comment by pavanbalaji** [comment]
I was referring to making the internal code more generic, but yes, the CVARs need to be more generic too.  What you suggested sounds reasonable.

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp

File: src/mpid/common/shm/mpidu_shm_alloc.c [hunk 2]
```diff
@@ -40,20 +58,29 @@ typedef struct alloc_elem {
...
-#define ALLOCQ_HEAD() GENERIC_Q_HEAD(allocq)
-#define ALLOCQ_EMPTY() GENERIC_Q_EMPTY(allocq)
-#define ALLOCQ_ENQUEUE(ep) GENERIC_Q_ENQUEUE(&allocq, ep, next)
-#define ALLOCQ_DEQUEUE(epp) GENERIC_Q_DEQUEUE(&allocq, epp, next)
-
 #define ROUND_UP_8(x) (((x) + (size_t)7) & ~(size_t)7)  /* rounds up to multiple of 8 */
 
-static size_t segment_len = 0;
+#ifdef HAVE_HWLOC
+#define SET_SHM_OBJ_BINDING(base_addr,len) do {                            \
```
Comment (gcongiu) [inline src/mpid/common/shm/mpidu_shm_alloc.c]: @pavanbalaji I hope I have addressed your point (I had to do some more invasive changes tough). I haven't tested this on knl yet. Waiting for JLSE to come back online.

File: src/mpi/init/initthread.c [hunk 3]
```diff
@@ -359,6 +359,45 @@ int MPIR_Init_thread(int *argc, char ***argv, int required, int *provided)
...
     }
+
+    /* detect HBM nodes */
+    hwloc_obj_t obj;
+    MPIR_Process.hbmset_is_valid = 0;
+    int i, numa_nodes = hwloc_get_nbobjs_by_type(MPIR_Process.hwloc_topology, HWLOC_OBJ_NUMANODE);
+    MPIR_Process.hbm_nodes = hwloc_bitmap_alloc();
+    for (i = 0; i < numa_nodes; i++) {
+        obj = hwloc_get_obj_by_type(MPIR_Process.hwloc_topology, HWLOC_OBJ_NUMANODE, i);
+        if (obj->subtype && strcmp(obj->subtype, "MCDRAM") == 0) {
```
Comment (pavanbalaji) [inline src/mpi/init/initthread.c]: Isn't this assuming that `HBM` is basically the same as `MCDRAM`?  It'd have been cleaner if you had a `hbm_type` that could default to `auto`, meaning whatever `hwloc` tells us is available on the node.  The user can force different placements when there are multiple other types of memories available.

In fact, it'd be even cleaner if we simply had `memory_type`.  `DRAM` is one memory type.  `MCDRAM` is another.  There might be more in the future.

File: src/mpid/ch3/channels/nemesis/src/mpid_nem_lmt_shm.c [hunk 4]
```diff
@@ -179,6 +179,24 @@ int MPID_nem_lmt_shm_start_recv(MPIDI_VC_t *vc, MPIR_Request *req, MPL_IOV s_coo
         mpi_errno = MPID_nem_allocate_shm_region(&vc_ch->lmt_copy_buf, vc_ch->lmt_copy_buf_handle);
         if (mpi_errno) MPIR_ERR_POP(mpi_errno);
 
+#ifdef HAVE_HWLOC
+        /* bind shared memory to HBM */
+        if (MPIR_Process.hbmset_is_valid) {
+            int info_flag;
+            char info_value[MPI_MAX_INFO_VAL+1];
+            MPIR_Info_get_impl(MPIR_Process.obj_mbind, "COPYBUFS", MPI_MAX_INFO_VAL, info_value,
```
Comment (pavanbalaji) [inline src/mpid/ch3/channels/nemesis/src/mpid_nem_lmt_shm.c]: We typically use small letters for info keys.  Also, as a separate cleanup pass, we should probably update all mpich-specific info keys to prefix them with `mpich_`.  Obviously not a part of this PR, but it'll be good to create an issue for that.

File: test/mpi/pt2pt/testlist.def [hunk 5]
```diff
@@ -48,3 +48,10 @@ dtype_send 2
 recv_any 2
 irecv_any 2
 large_tag 2
+hbm 2 arg=-membind=FASTBOXES:HBM,CELLS:DRAM,COPYBUFS:DRAM
+hbm 2 arg=-membind=FASTBOXES:DRAM,CELLS:HBM,COPYBUFS:DRAM
+hbm 2 arg=-membind=FASTBOXES:DRAM,CELLS:DRAM,COPYBUFS:HBM
+hbm 2 arg=-membind=FASTBOXES:HBM,CELLS:HBM,COPYBUFS:DRAM
+hbm 2 arg=-membind=FASTBOXES:HBM,CELLS:DRAM,COPYBUFS:HBM
+hbm 2 arg=-membind=FASTBOXES:HBM,CELLS:HBM,COPYBUFS:HBM
+hbm 2 arg=-membind=FASTBOXES:DRAM,CELLS:HBM,COPYBUFS:HBM
```
Comment (pavanbalaji) [inline test/mpi/pt2pt/testlist.def]: Do we need to specify `DRAM`?  Isn't that the default?  Or are you simply doing it for completeness?

File: src/include/mpir_objects.h [hunk 6]
```diff
@@ -159,6 +159,10 @@ typedef enum MPII_Object_kind {
     MPIR_GREQ_CLASS = 0xf
 } MPII_Object_kind;
 
+typedef enum MPII_Memory_kind {
+    MPIR_MEMKIND__DRAM = 0,
+    MPIR_MEMKIND__HBM
+} MPII_Memory_kind;
```
[the remark this answers is not in the document]
Comment (gcongiu) [reply src/include/mpir_objects.h]: Yes the default is DRAM, no need to specify membind list correct.

File: src/mpi/init/initthread.c (hunk 3 above)
Comment (gcongiu) [reply src/mpi/init/initthread.c]: This is specific to MCDRAM since we didn't have other devices we could play with. That `if(obj->subtype && strcmp(obj->subtype, "MCDRAM")` can be extended with multiple `else if` to include other memories.

File: src/mpid/ch3/channels/nemesis/src/mpid_nem_lmt_shm.c (hunk 4 above)
Comment (gcongiu) [reply src/mpid/ch3/channels/nemesis/src/mpid_nem_lmt_shm.c]: I will convert it to lower case and create an issue for the rest of the info keys.

File: test/mpi/pt2pt/testlist.def (hunk 5 above)
Comment (gcongiu) [reply test/mpi/pt2pt/testlist.def]: Yes just for completeness

File: src/mpi/init/initthread.c (hunk 3 above)
Comment (pavanbalaji) [reply src/mpi/init/initthread.c]: I'm not asking you to test with more memories.  I'm just saying that your current code assumes that HBM == MCDRAM.  I think the detection code should be separated from the usage code.  For detection, we can detect all available forms of memory: DDR, MCDRAM, ...  For usage, we should default to DDR and if the user asks for something else, then we can see if such memory is available and use it.

Comment (gcongiu) [reply src/mpi/init/initthread.c]: Ok I will think about how to integrate the changes to address you comment

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp

File: src/include/mpir_memkind.h [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+#ifndef MPIR_MEMKIND_H_INCLUDED
+#define MPIR_MEMKIND_H_INCLUDED
+
+#include "mpichconf.h"
+
+typedef enum MPII_Memory_kind {
+    MPIR_MEMKIND__MCDRAM = 0,
+    /* Future memory types must be added after MCDRAM */
+    MPIR_MEMKIND__NUM,
```
Comment (pavanbalaji) [inline src/include/mpir_memkind.h]: What is `NUM`?  If you want the number of memory types, why isn't `DRAM` in that list?

File: src/mpi/init/initthread.c [hunk 8]
```diff
@@ -31,6 +31,8 @@
     - name        : DEBUGGER
       description : cvars relevant to the "MPIR" debugger interface
 
+    - name        : MEMBIND
+      description : cvars controlling shared memory objects binding
```
Comment (pavanbalaji) [inline src/mpi/init/initthread.c]: Can you add a new line after the categories?

File: src/mpi/init/initthread.c [hunk 9]
```diff
@@ -68,6 +70,16 @@
...
+    - name        : MPIR_CVAR_OBJECT_MEMBIND_LIST
+      category    : MEMBIND
+      type        : string
+      default     : "FASTBOXES:AUTO,CELLS:AUTO,COPYBUFS:AUTO"
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        The memory types used for allocating specific MPICH objects in a
+        heterogeneous memory system.
```
Comment (pavanbalaji) [inline src/mpi/init/initthread.c]: Newline before the end of the block.

File: src/mpi/init/initthread.c [hunk 10]
```diff
@@ -359,6 +374,81 @@ int MPIR_Init_thread(int *argc, char ***argv, int required, int *provided)
...
+        MPIR_Process.memkind_is_avail[i] = 0;
+    }
+
+    for (i = 0; i < numa_nodes; i++) {
+        obj = hwloc_get_obj_by_type(MPIR_Process.hwloc_topology, HWLOC_OBJ_NUMANODE, i);
+        if (obj->subtype && strcmp(obj->subtype, "MCDRAM") == 0) {
+            hwloc_bitmap_or(MPIR_Process.memkind_nodes[MPIR_MEMKIND__MCDRAM], obj->nodeset,
+                            MPIR_Process.memkind_nodes[MPIR_MEMKIND__MCDRAM]);
+        }
+    }
```
Comment (pavanbalaji) [inline src/mpi/init/initthread.c]: I think it'd be cleaner to do this for all memory kinds, including DRAM.

File: src/mpid/ch3/channels/nemesis/src/mpid_nem_init.c [hunk 11]
```diff
@@ -238,24 +241,42 @@ MPID_nem_init(int pg_rank, MPIDI_PG_t *pg_p, int has_parent ATTRIBUTE((unused)))
 #endif  /*FORCE_ASYM */
 
     /* Request fastboxes region */
-    mpi_errno = MPIDU_shm_seg_alloc(MPL_MAX((num_local*((num_local-1)*sizeof(MPID_nem_fastbox_t))), MPID_NEM_ASYMM_NULL_VAL),
-                                     (void **)&fastboxes_p, MPL_MEM_SHM);
+#ifdef HAVE_HWLOC
+    MPIR_Info_get_impl(MPIR_Process.obj_mbind, "fastboxes", MPI_MAX_INFO_VAL, info_value, &info_flag);
+    if (info_flag)
+        kind = (strcmp(info_value, "mcdram") == 0 && MPIR_Process.memkind_is_avail[MPIR_MEMKIND__MCDRAM]) ?
+            MPIR_MEMKIND__MCDRAM : MPIR_MEMKIND__DRAM;
```
Comment (pavanbalaji) [inline src/mpid/ch3/channels/nemesis/src/mpid_nem_init.c]: Can you add an error case if the user picks an unsupported value, e.g., `foobar`?  Also, if the user picked `mcdram` and there's no `MCDRAM` available, you are silently falling back to `DRAM`.  Is that the right thing to do or should we throw an error?  I'm not sure what the answer is.  Maybe others in the group have an opinion.

Btw, on what is this info set?  Since fastbox/cells are not specific to a communicator or a window, I'm not sure info keys make too much sense here.  I'm sorry if I was the one who suggested doing that in the first place.  It sounds like that was bad advice.  :-)

File: src/mpid/ch3/channels/nemesis/src/mpid_nem_init.c [hunk 12]
```diff
@@ -238,24 +241,42 @@ MPID_nem_init(int pg_rank, MPIDI_PG_t *pg_p, int has_parent ATTRIBUTE((unused)))
...
+                                    MPL_MEM_SHM);
     if (mpi_errno) MPIR_ERR_POP(mpi_errno);
     
     /* Request data cells region */
-    mpi_errno = MPIDU_shm_seg_alloc(num_local * MPID_NEM_NUM_CELLS * sizeof(MPID_nem_cell_t), (void **)&cells_p, MPL_MEM_SHM);
+#ifdef HAVE_HWLOC
+    MPIR_Info_get_impl(MPIR_Process.obj_mbind, "cells", MPI_MAX_INFO_VAL, info_value, &info_flag);
+    if (info_flag)
+        kind = (strcmp(info_value, "mcdram") == 0 && MPIR_Process.memkind_is_avail[MPIR_MEMKIND__MCDRAM]) ?
+            MPIR_MEMKIND__MCDRAM : MPIR_MEMKIND__DRAM;
```
Comment (pavanbalaji) [inline src/mpid/ch3/channels/nemesis/src/mpid_nem_init.c]: Same comment as above.

File: src/mpid/ch4/shm/posix/posix_init.h [hunk 13]
```diff
@@ -78,39 +78,41 @@ static inline int MPIDI_POSIX_mpi_init_hook(int rank, int size, int *n_vnis_prov
     mpi_errno =
         MPIDU_shm_seg_alloc(MAX
                             ((num_local * ((num_local - 1) * sizeof(MPIDI_POSIX_fastbox_t))),
-                             MPIDI_POSIX_ASYMM_NULL_VAL), (void **) &fastboxes_p, MPL_MEM_SHM);
+                             MPIDI_POSIX_ASYMM_NULL_VAL), MPIR_MEMKIND__DRAM,
+                            (void **) &fastboxes_p, MPL_MEM_SHM);
```
Comment (pavanbalaji) [inline src/mpid/ch4/shm/posix/posix_init.h]: I'm probably missing some context here.  Is this being forced to `DRAM` instead of checking for the user preference?  This happens more times below.  Some of them are non-cell/fastbox allocations, which are fine.  But thought we are allowing for multiple memory types for fbox and cells.

File: src/mpid/common/shm/mpidu_shm_alloc.c [hunk 14]
```diff
@@ -36,24 +38,34 @@ typedef struct alloc_elem {
...
                        int num_local, int local_rank);
 
-#define ALLOCQ_HEAD() GENERIC_Q_HEAD(allocq)
-#define ALLOCQ_EMPTY() GENERIC_Q_EMPTY(allocq)
-#define ALLOCQ_ENQUEUE(ep) GENERIC_Q_ENQUEUE(&allocq, ep, next)
-#define ALLOCQ_DEQUEUE(epp) GENERIC_Q_DEQUEUE(&allocq, epp, next)
-
 #define ROUND_UP_8(x) (((x) + (size_t)7) & ~(size_t)7)  /* rounds up to multiple of 8 */
+#ifdef HAVE_HWLOC
+#define MPII_OBJECT_MEMDIND(base_addr,len,kind) do {                       \
```
Comment (pavanbalaji) [inline src/mpid/common/shm/mpidu_shm_alloc.c]: Typo in the macro name?

File: src/mpid/common/shm/mpidu_shm_alloc.c [hunk 15]
```diff
@@ -100,10 +112,15 @@ int MPIDU_shm_seg_alloc(size_t len, void **ptr_p, MPL_memory_class class)
...
-    ALLOCQ_ENQUEUE(ep);
-
-    segment_len += len;
+    if (kind == MPIR_MEMKIND__DRAM) {
+        GENERIC_Q_ENQUEUE(&allocq_dram, ep, next);
+        dram_segment_len += len;
+    } else {
+        GENERIC_Q_ENQUEUE(&allocq_hybr, ep, next);
+        hybr_segment_len += len;
+    }
```
Comment (pavanbalaji) [inline src/mpid/common/shm/mpidu_shm_alloc.c]: I'm not missing some logic here.  Once the memory is allocated in either `DRAM` or `MCDRAM`, why does the rest of the code need to know about it?

File: src/mpid/ch3/src/ch3u_win_fns.c [hunk 16]
```diff
@@ -544,3 +592,66 @@ int MPID_Win_get_info(MPIR_Win * win, MPIR_Info ** info_used)
   fn_fail:
     goto fn_exit;
 }
+
+#undef FUNCNAME
+#define FUNCNAME win_membind_set
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+void win_membind_set(MPIR_Info * info, void *baseptr, size_t len)
```
Comment (pavanbalaji) [inline src/mpid/ch3/src/ch3u_win_fns.c]: Isn't this a `static` function?

File: src/mpid/ch3/src/ch3u_win_fns.c [hunk 17]
```diff
@@ -7,9 +7,56 @@
 #include "mpidimpl.h"
 #include "mpir_info.h"
 #include "mpidrma.h"
+#include "mpir_memkind.h"
```
Comment (pavanbalaji) [inline src/mpid/ch3/src/ch3u_win_fns.c]: Are we adding this support for `ch3` too, or just `ch4`?

Comment (gcongiu) [reply src/mpid/ch3/src/ch3u_win_fns.c]: CH4 was not supporting fastboxes at the time I started this. I will add it now

File: src/mpid/common/shm/mpidu_shm_alloc.c (hunk 14 above)
Comment (gcongiu) [reply src/mpid/common/shm/mpidu_shm_alloc.c]: yes

File: src/mpid/ch3/channels/nemesis/src/mpid_nem_init.c (hunk 11 above)
Comment (gcongiu) [reply src/mpid/ch3/channels/nemesis/src/mpid_nem_init.c]: I am not familiar with the error handling inside MPICH can you provide some hint on how this should be done. I suspect I need to create an error code and then pass it to upper layers with `MPIR_ERR_POP`?

Comment (pavanbalaji) [reply src/mpid/ch3/channels/nemesis/src/mpid_nem_init.c]: We should probably discuss what object this info is set on first.  Because, if that's the wrong thing to do, you can probably entirely delete this part of the code.

Comment (gcongiu) [reply src/mpid/ch3/channels/nemesis/src/mpid_nem_init.c]: Sure

File: src/mpid/ch4/shm/posix/posix_init.h (hunk 13 above)
Comment (gcongiu) [reply src/mpid/ch4/shm/posix/posix_init.h]: This should be addressed now

File: src/mpid/common/shm/mpidu_shm_alloc.c (hunk 15 above)
Comment (gcongiu) [reply src/mpid/common/shm/mpidu_shm_alloc.c]: After memory has been accounted by `MPIDU_shm_seg_alloc`s it has to be allocated and bound to the corresponding memory type in `MPIDU_shm_seg_commit`. The `allocq_hybr` has to retain the memory type information (`MCDRAM`, `NVRAM`, etc) so that the commit function can perform the right binding. This does not allow for mixed hybrid memory allocations (e.g., `FASTBOXES:MCDRAM` and `CELLS:NVRAM`).

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
This PR has dependencies with  #3490

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
@raffenet I am having segfaults on some RMA tests for CH4/OFI. I am not sure this is related to my patch since I have re-run the tests manually on the same bb nodes and they succeed.

**Comment by pavanbalaji** [comment]
`mpich/master` is clean.  There are no such segfaults.

**Comment by raffenet** [comment]
@gcongiu make sure to use the odd/even configuration for accurate reproduction. `MPIR_CVAR_ODD_EVEN_CLIQUES=1`

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
@raffenet @pavanbalaji CH4 tests now complete correctly (for some reason jenkins is failing to collect the summary and thus is not showing tests as passed). There are still a few corrections I need to make to the code but before doing it I would like to get your comments.

**Comment by pavanbalaji** [comment]
This exposes fastboxes and copy buffers to the MPI layer.  The abstraction needs to be improved.

**Comment by gcongiu** [comment]
> This exposes fastboxes and copy buffers to the MPI layer. The abstraction needs to be improved.

Should I move `mpir_memkind.h` to the common device layer instead (i.e., `mpidu_memkind.h`)? Concerning `MPIR_Process` is it ok to have additional `numa_info` data in there? Or should this be also moved to the device layer and reside in another global data structure?

**Comment by pavanbalaji** [comment]
> Should I move `mpir_memkind.h` to the common device layer instead (i.e., `mpidu_memkind.h`)?

Not everything is device specific.  So you might need to split it.

> Concerning `MPIR_Process` is it ok to have additional `numa_info` data in there? Or should this be also moved to the device layer and reside in another global data structure?

Why not generalize it to expose topology in a more uniform fashion instead of hardcoding just "node" and "numa"?  We now use `hwloc` at the MPI layer, so we might as well use those objects.

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
Running additional tests as OFI works. @raffenet I still haven't figure out why report collection is failing. 
test:jenkins/ch4/ucx
test:jenkins/ch3/tcp

File: test/mpi/pt2pt/testlist.def [hunk 18]
```diff
@@ -50,3 +50,11 @@ dtype_send 2
 recv_any 2
 irecv_any 2
 large_tag 2
+
+# Heterogeneous memory tests
+sendflood 8 env=MPIR_CVAR_MEMBIND_NUMA_ENABLE="YES" env=MPIR_CVAR_MEMBIND_TYPE_LIST="FASTBOXES:AUTO" env=MPIR_CVAR_MEMBIND_POLICY_LIST="FASTBOXES:BIND" env=MPIR_CVAR_MEMBIND_FLAGS_LIST="FASTBOXES:STRICT" timeLimit=600
```
Comment (raffenet) [inline test/mpi/pt2pt/testlist.def]: It looks like these additional quotes are causing the XML parser to barf. Can you remove them? E.g. ```sendflood 8 env=MPIR_CVAR_MEMBIND_NUMA_ENABLE=YES env=MPIR_CVAR_MEMBIND_TYPE_LIST=FASTBOXES:AUTO env=MPIR_CVAR_MEMBIND_POLICY_LIST=FASTBOXES:BIND env=MPIR_CVAR_MEMBIND_FLAGS_LIST=FASTBOXES:STRICT timeLimit=600```

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/most

**Comment by gcongiu** [comment]
@pavanbalaji @raffenet tests seem fine. I noticed a few minor things that I need to fix. However, if you want you can start looking at the code.

File: src/include/mpir_memtype.h [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+
+#ifndef MPIR_MEMTYPE_H_INCLUDED
+#define MPIR_MEMTYPE_H_INCLUDED
+
+typedef enum MPIR_Memtype {
+    MPIR_MEMTYPE__NONE = -1,
+    MPIR_MEMTYPE__DDR = 0,
+    MPIR_MEMTYPE__MCDRAM,
+    MPIR_MEMTYPE__NUM,
+    MPIR_MEMTYPE__DEFAULT = MPIR_MEMTYPE__DDR
```
Comment (pavanbalaji) [inline src/include/mpir_memtype.h]: Is that allowed in C99?

File: src/mpid/common/shm/mpidu_shm_obj.h [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+#define MPIDU_SHM_OBJ_H_INCLUDED
+
+typedef enum MPIDU_shm_obj {
+    MPIDU_SHM_OBJ__NONE = -1,
+    MPIDU_SHM_OBJ__FASTBOXES = 0,
+    MPIDU_SHM_OBJ__CELLS,
+    MPIDU_SHM_OBJ__COPYBUFS,
+    MPIDU_SHM_OBJ__WIN,
+    MPIDU_SHM_OBJ__NUM
+} MPIDU_shm_obj_t;
```
Comment (pavanbalaji) [inline src/mpid/common/shm/mpidu_shm_obj.h]: The memory objects (I'd call these `buffer types`, instead) are different from memory types.  These two should be split into different commits.  Memory types are usable even without explicit buffer types, i.e., we can have all buffers be allocated on DRAM or MCDRAM or something else.  Buffer types only allow us to have some buffers on one memory type and other buffers on a different memory type.

File: src/include/mpir_memtype.h (hunk 19 above)
Comment (gcongiu) [reply src/include/mpir_memtype.h]: It should be allowed. The C99 standard does not say you can't do it (`section 6.7.2.2`):
`The expression that defines the value of an enumeration constant shall be an integer
constant expression that has a value representable as an int. [...] (The use of enumerators with
= may produce enumeration constants with values that duplicate other values in the same
enumeration.)`. My interpretation might be wrong though.

File: src/mpid/common/shm/mpidu_shm_obj.h (hunk 20 above)
Comment (gcongiu) [reply src/mpid/common/shm/mpidu_shm_obj.h]: Makes sense, I will split the two in different commits.

**Comment by gcongiu** [comment]
I am closing this PR as the code has to be completely rewritten and is cleaner to start over with another one.
